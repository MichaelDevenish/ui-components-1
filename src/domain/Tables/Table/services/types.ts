import { IFontAwesomeIconButtonProps } from '../../../../domain/Buttons/FontAwesomeIconButton/FontAwesomeIconButton'
import { TableRowVariant } from './colors'

enum ColumnAlignment {
  Left = 'left',
  Right = 'right'
}

enum ColumnSize {
  Auto = 'auto',
  Shrink = 'shrink'
}
enum ColumnSortDirection {
  Descending = 'descending',
  Ascending = 'ascending'
}

enum InteractionType {
  Hover = 'hover',
  Swipe = 'swipe'
}

enum TableCheckboxInputValue {
  True = 'true',
  False = 'false',
  PartialTrue = 'partialTrue'
}

interface ISelectedRows {
  [rowId: string]: boolean
}

interface IColumnSorts {
  [columnName: string]: ColumnSortDirection
}

interface IRowProps <T> {
  id: string
  isSelectable?: boolean
  isRemovable?: boolean
  tooltipText?: string
  progress?: number
  variant?: TableRowVariant
  data: T
  contentOverride?: (rowData: T) => JSX.Element[] | JSX.Element
  onClick?: (rowData: T) => void
  actions?: IFontAwesomeIconButtonProps[]
}

interface IColumnProps <T> {
  name: string
  title?: string
  size: ColumnSize
  headerSize?: ColumnSize
  content: (rowData: T) => JSX.Element
  alignment?: ColumnAlignment
  tooltipText?: (rowData: T) => string
}

export {
  ColumnAlignment,
  ColumnSize,
  ColumnSortDirection,
  InteractionType,
  TableCheckboxInputValue,
  ISelectedRows,
  IColumnSorts,
  IRowProps,
  IColumnProps
}
