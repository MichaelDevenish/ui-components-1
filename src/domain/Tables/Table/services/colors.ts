import { Variables } from '../../../../common'

enum RowVariant {
  Neutral = 'neutral',
  Red = 'red'
}

interface IVariantOption {
  hoverBackground: Variables.Color
  selectBackground: Variables.Color
}

type VariantOptions = {
  [K in RowVariant]: IVariantOption
}

const variantOptions: VariantOptions = {
  [RowVariant.Neutral]: {
    hoverBackground: Variables.Color.n150,
    selectBackground: Variables.Color.i100
  },
  [RowVariant.Red]: {
    hoverBackground: Variables.Color.r100,
    selectBackground: Variables.Color.r100
  }
}

export {
  variantOptions,
  RowVariant
}
