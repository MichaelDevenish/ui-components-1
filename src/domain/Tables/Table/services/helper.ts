import { useEffect, useRef } from 'react'

const usePrevious = <T extends {}>(value: T): T => {
  const ref = useRef<T>()
  useEffect(() => {
    ref.current = value
  }, [value])

  return ref.current as T
}

const handleHovered = (value: boolean, setHasHovered: (value: boolean) => void) => () => setHasHovered(value)

export {
  usePrevious,
  handleHovered
}
