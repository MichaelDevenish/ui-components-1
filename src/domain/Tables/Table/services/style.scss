:local(.table)  {
  .row-enter {
    opacity: 0;
  }
  .row-enter-active {
    opacity: 1;
    transition: opacity 500ms ease-in;
  }
  .row-exit {
    opacity: 1;
  }
  .row-exit-active {
    opacity: 0;
    transition: opacity 500ms ease-in;
  }
}
