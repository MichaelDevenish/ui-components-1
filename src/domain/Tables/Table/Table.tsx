import { isEqual } from 'lodash'
import React, { useCallback, useEffect, useMemo, useState } from 'react'

import { Props } from '../../../common'
import {
  IFontAwesomeIconButtonProps
} from '../../Buttons/FontAwesomeIconButton/FontAwesomeIconButton'
import { TableRowVariant } from './services/colors'
import {
  getSelectAllTableCheckboxInputValue,
  getUpdatedAllSelectableRows,
  handleSelectionChanged
} from './services/helper'
import {
  StyledEmptyStateCell,
  StyledEmptyStateRow,
  StyledTHead,
  StyledTable,
  StyledTableWrapper
} from './services/style'
import {
  ColumnAlignment,
  ColumnSize,
  ColumnSortDirection,
  IColumnProps,
  IColumnSorts,
  IRowProps,
  ISelectedRows,
  InteractionType
} from './services/types'
import { TableCheckboxInputValue } from './subcomponents/TableCheckboxInput'
import { TableHeader } from './subcomponents/TableHeader'
import { TableRow } from './subcomponents/TableRow'

const validateProps = <T extends {}>(props: ITableProps<T>) => {
  if (props.hasLeftAction && props.interactionType === InteractionType.Swipe) {
    throw new Error('Having left actions is not compatible with swipe interaction')
  }

  if (props.onSortChange && !props.sort) {
    throw new Error('You must provide a sort for onSortChange to work')
  }
}

interface ITableProps <T extends {}> {
  /** Rows of the table */
  rows: Array<IRowProps<T>>
  /** Columns of the table */
  columns: Array<IColumnProps<T>>
  /** Empty state of the table */
  emptyState: JSX.Element
  /** Sorting of the table and only allow one column sorting at once */
  sort?: IColumnSorts
  /** Called when the sorting is changed */
  onSortChange?: (sort: IColumnSorts) => void
  /** Called when the selected row is changed */
  onSelectedRowChange?: (selectedRowData: T[]) => void
  /** Icon Buttons group for bulkActions when the user selects some row */
  bulkActions?: IFontAwesomeIconButtonProps[]
  /** Called when some row is removed */
  onRowRemove?: (removedRowData: T) => void
  /** If yes the left action cells are displayed (checkbox and remove button) */
  hasLeftAction?: boolean
  /** If interaction type is Hover, it will display hover actions when hover. If interaction type is swipe, it will display swipe actions when tap or swipe. */
  interactionType?: InteractionType
  /** If yes, sort is enable with interaction. If no, sort button display in disable style with given sort */
  hasSortEnabled?: boolean
  /** Margins around the table */
  margins?: Props.IMargins
  /** The component context */
  componentContext?: string
}

const Table = <T extends {}>(props: ITableProps<T>) => {
  const {
    onRowRemove,
    rows,
    sort,
    onSortChange,
    columns,
    onSelectedRowChange,
    margins,
    componentContext,
    hasLeftAction = false,
    bulkActions,
    emptyState,
    interactionType = InteractionType.Hover,
    hasSortEnabled = true
  } = props

  validateProps(props)
  const [selectedRows, setSelectedRows] = useState<ISelectedRows>(() => getUpdatedAllSelectableRows(rows, {}))
  const [expandedSwipeCellRow, setExpandedSwipeCellRow] = useState<string | null>(null)

  const selectAllTableCheckboxInputValue = useMemo(() => getSelectAllTableCheckboxInputValue(Object.values(selectedRows)), [selectedRows])
  const areRowsSelected = useMemo(() => {
    return Object.values(selectedRows).includes(true)
  }, [selectedRows])

  const setAllRows = useCallback((value: TableCheckboxInputValue) => {
    const newSelectedRows = rows.reduce<ISelectedRows>((acc, row) => {
      acc[row.id] = (value === TableCheckboxInputValue.True)

      return acc
    }, {})

    setSelectedRows(newSelectedRows)
  }, [rows, setSelectedRows])

  const toggleSingleRow = useCallback((row: IRowProps<T>) => {
    setSelectedRows(
      (selectedRowsFromState) => ({...selectedRowsFromState, [row.id]: !selectedRowsFromState[row.id] })
    )
  }, [setSelectedRows])

  useEffect(() => {
    const updatedSelectedRows = getUpdatedAllSelectableRows<T>(rows, selectedRows)

    if (!isEqual(updatedSelectedRows, selectedRows)) {
      setSelectedRows(updatedSelectedRows)
    }
  }, [rows])
  useEffect(() => {
    if (hasLeftAction) {
      setSelectedRows({})
    }
  }, [hasLeftAction])
  useEffect(() => {
    if (hasLeftAction && onSelectedRowChange) {
      handleSelectionChanged<T>(rows, selectedRows, onSelectedRowChange)
    }
  }, [selectedRows])

  const hasTableSwipeActions = interactionType === InteractionType.Swipe && rows.some((row) => !!row.actions && row.actions.length > 0)

  return (
    <StyledTableWrapper
      margins={margins}
      data-component-type={Props.ComponentType.Table}
      data-component-context={componentContext}
    >
      <StyledTable>
        <StyledTHead>
          <TableHeader<T>
            sort={sort}
            onSortChange={onSortChange}
            columns={columns}
            selectAllTableCheckboxInputValue={selectAllTableCheckboxInputValue}
            setSelectAllTableCheckboxInputValue={setAllRows}
            hasLeftAction={hasLeftAction}
            bulkActions={bulkActions}
            hasTableSwipeActions={hasTableSwipeActions}
            hasBulkAction={selectAllTableCheckboxInputValue !== TableCheckboxInputValue.False}
            isEmpty={rows.length === 0}
            hasSortEnabled={hasSortEnabled}
          />
        </StyledTHead>
          {
            rows.length === 0 ? (
              <StyledEmptyStateRow>
                <StyledEmptyStateCell colSpan={columns.length}>
                  {emptyState}
                </StyledEmptyStateCell>
              </StyledEmptyStateRow>
            ) : rows.map((row, index) => (
              <TableRow<T>
                key={row.id}
                row={row}
                lastRow={index === rows.length - 1}
                hasTableSwipeActions={hasTableSwipeActions}
                columns={columns}
                areRowsSelected={areRowsSelected}
                isSelected={selectedRows[row.id] || false}
                toggleSelected={toggleSingleRow}
                hasLeftAction={hasLeftAction}
                interactionType={interactionType}
                onRowRemove={onRowRemove}
                expandedSwipeCellRow={expandedSwipeCellRow}
                setExpandedSwipeCellRow={setExpandedSwipeCellRow}
              />
            ))
          }
      </StyledTable>
    </StyledTableWrapper>
  )
}

type ITableType = typeof Table & {
  ColumnAlignment: typeof ColumnAlignment
  ColumnSize: typeof ColumnSize
  ColumnSortDirection: typeof ColumnSortDirection
  InteractionType: typeof InteractionType
  RowVariant: typeof TableRowVariant
}

const MemoTable: ITableType = React.memo(Table) as any

MemoTable.ColumnSize = ColumnSize
MemoTable.ColumnAlignment = ColumnAlignment
MemoTable.ColumnSortDirection = ColumnSortDirection
MemoTable.InteractionType = InteractionType
MemoTable.RowVariant = TableRowVariant

export {
  Table as UnmemoTable, // Needed for styleguide (memoed components won't be detected)
  MemoTable as Table
}
