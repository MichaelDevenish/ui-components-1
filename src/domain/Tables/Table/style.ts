import styled from 'styled-components'
import {styleForMargins} from '../../Spacers/services/margins'
import {Props, Variables} from '../../../common'

interface IStyledTableProps {
  margins?: Props.IMargins
}

const StyledTable = styled.div`
  margin: 0;
  border: 1px solid ${Variables.Color.n250};

  ${(props: IStyledTableProps) => styleForMargins(props.margins)}
`

export {
  StyledTable
}
