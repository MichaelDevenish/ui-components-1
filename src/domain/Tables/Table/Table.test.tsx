import { shallow } from 'enzyme'
import React from 'react'

import {TableRowVariant} from './services/colors'
import {ColumnAlignment, ColumnSize, ColumnSortDirection, Table, TableTouchType} from './Table'

const onSelectionChanged = jest.fn()
const onSortChange = jest.fn()
const onRowClick = jest.fn()
const emptyState = <div>empty</div>
const contentOverride = () => [<div key={0}>override</div>, <div key={1}>override</div>]
const rowContentOverride = () => <div>override</div>
const columns = [
  {
    name: 'fileName',
    title: 'file name',
    size: ColumnSize.Auto,
    content: (data: any) => <div>{data.fileName}</div>,
    alignment: ColumnAlignment.Left
  },
  {
    name: 'createAt',
    title: 'created at',
    size: ColumnSize.Shrink,
    content: (data: any) => <div>{data.createAt}</div>,
    alignment: ColumnAlignment.Right
  }
]
const rows = [
  {
    id: '0',
    data: {fileName: 'fail.pdf', createAt: '05/01/2018', size: '8.2', fileType: 'PDF'},
    isSelectable: false,
    isRemovable: true,
    variant: TableRowVariant.Red,
    onClick: onRowClick,
    rowContentOverride
  },
  {
    id: '1',
    data: {fileName: 'uploading.pdf', createAt: '05/01/2018', size: '99', fileType: 'PDF'},
    isSelectable: false,
    isRemovable: true,
    progress: 0.7,
    contentOverride
  },
  {
    id: '2',
    data: {fileName: 'success.pdf', createAt: '05/01/2018', size: '3.7', fileType: 'PDF'},
    isSelectable: false,
    isRemovable: true,
    progress: 0.5,
    contentOverride
  },
  {
    id: '3',
    data: {fileName: 'test.pdf', createAt: '05/01/2018', size: '2.2', fileType: 'PDF'},
    isSelectable: true
  },
  {
    id: '4',
    data: {fileName: 'test1.pdf', createAt: '05/01/2018', size: '1.5', fileType: 'PDF'},
    isSelectable: true
  }
]

describe('<Table />', () => {
  it('should render a table', () => {
    const wrapper = shallow(
      <Table
        hasLeftAction
        onSelectionChanged={onSelectionChanged}
        rows={rows}
        sort={{createAt: ColumnSortDirection.Down}}
        onSortChange={onSortChange}
        emptyState={emptyState}
        columns={columns}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a mobile table', () => {
    const wrapper = shallow(
      <Table
        touchType={TableTouchType.Swipe}
        onSelectionChanged={onSelectionChanged}
        rows={rows}
        sort={{createAt: ColumnSortDirection.Down}}
        onSortChange={onSortChange}
        emptyState={emptyState}
        columns={columns}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render an empty table', () => {
    const wrapper = shallow(
      <Table
        onSelectionChanged={onSelectionChanged}
        rows={[]}
        sort={{createAt: ColumnSortDirection.Down}}
        onSortChange={onSortChange}
        emptyState={emptyState}
        columns={columns}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
