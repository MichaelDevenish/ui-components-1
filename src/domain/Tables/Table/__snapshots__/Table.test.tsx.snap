// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render a mobile table 1`] = `
<styled.div
  data-component-type="table"
>
  <styled.table>
    <styled.thead>
      <Memo(TableHeader)
        columns={
          Array [
            Object {
              "alignment": "left",
              "content": [Function],
              "name": "fileName",
              "size": "auto",
              "title": "file name",
            },
            Object {
              "alignment": "right",
              "content": [Function],
              "name": "createAt",
              "size": "shrink",
              "title": "created at",
            },
          ]
        }
        hasBulkAction={false}
        hasLeftAction={false}
        hasSortEnabled={true}
        hasTableSwipeActions={true}
        isEmpty={false}
        onSortChange={[MockFunction]}
        selectAllTableCheckboxInputValue="false"
        setSelectAllTableCheckboxInputValue={[Function]}
        sort={
          Object {
            "createAt": "ascending",
          }
        }
      />
    </styled.thead>
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={false}
      hasTableSwipeActions={true}
      interactionType="swipe"
      isSelected={false}
      key="0"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "fail.pdf",
            "fileType": "PDF",
            "size": "8.2",
          },
          "id": "0",
          "isRemovable": true,
          "isSelectable": false,
          "onClick": [MockFunction],
          "rowContentOverride": [Function],
          "variant": "error",
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={false}
      hasTableSwipeActions={true}
      interactionType="swipe"
      isSelected={false}
      key="1"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "contentOverride": [Function],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "uploading.pdf",
            "fileType": "PDF",
            "size": "99",
          },
          "id": "1",
          "isRemovable": true,
          "isSelectable": false,
          "progress": 0.7,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={false}
      hasTableSwipeActions={true}
      interactionType="swipe"
      isSelected={false}
      key="2"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "contentOverride": [Function],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "success.pdf",
            "fileType": "PDF",
            "size": "3.7",
          },
          "id": "2",
          "isRemovable": true,
          "isSelectable": false,
          "progress": 0.5,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={false}
      hasTableSwipeActions={true}
      interactionType="swipe"
      isSelected={false}
      key="3"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "test.pdf",
            "fileType": "PDF",
            "size": "2.2",
          },
          "id": "3",
          "isSelectable": true,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={false}
      hasTableSwipeActions={true}
      interactionType="swipe"
      isSelected={false}
      key="4"
      lastRow={true}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "test1.pdf",
            "fileType": "PDF",
            "size": "1.5",
          },
          "id": "4",
          "isSelectable": true,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
  </styled.table>
</styled.div>
`;

exports[`<Table /> should render a table 1`] = `
<styled.div
  data-component-type="table"
>
  <styled.table>
    <styled.thead>
      <Memo(TableHeader)
        columns={
          Array [
            Object {
              "alignment": "left",
              "content": [Function],
              "name": "fileName",
              "size": "auto",
              "title": "file name",
            },
            Object {
              "alignment": "right",
              "content": [Function],
              "name": "createAt",
              "size": "shrink",
              "title": "created at",
            },
          ]
        }
        hasBulkAction={false}
        hasLeftAction={true}
        hasSortEnabled={true}
        hasTableSwipeActions={false}
        isEmpty={false}
        onSortChange={[MockFunction]}
        selectAllTableCheckboxInputValue="false"
        setSelectAllTableCheckboxInputValue={[Function]}
        sort={
          Object {
            "createAt": "ascending",
          }
        }
      />
    </styled.thead>
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={true}
      hasTableSwipeActions={false}
      interactionType="hover"
      isSelected={false}
      key="0"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "fail.pdf",
            "fileType": "PDF",
            "size": "8.2",
          },
          "id": "0",
          "isRemovable": true,
          "isSelectable": false,
          "onClick": [MockFunction],
          "rowContentOverride": [Function],
          "variant": "error",
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={true}
      hasTableSwipeActions={false}
      interactionType="hover"
      isSelected={false}
      key="1"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "contentOverride": [Function],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "uploading.pdf",
            "fileType": "PDF",
            "size": "99",
          },
          "id": "1",
          "isRemovable": true,
          "isSelectable": false,
          "progress": 0.7,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={true}
      hasTableSwipeActions={false}
      interactionType="hover"
      isSelected={false}
      key="2"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "contentOverride": [Function],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "success.pdf",
            "fileType": "PDF",
            "size": "3.7",
          },
          "id": "2",
          "isRemovable": true,
          "isSelectable": false,
          "progress": 0.5,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={true}
      hasTableSwipeActions={false}
      interactionType="hover"
      isSelected={false}
      key="3"
      lastRow={false}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "test.pdf",
            "fileType": "PDF",
            "size": "2.2",
          },
          "id": "3",
          "isSelectable": true,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
    <Memo(TableRow)
      areRowsSelected={false}
      columns={
        Array [
          Object {
            "alignment": "left",
            "content": [Function],
            "name": "fileName",
            "size": "auto",
            "title": "file name",
          },
          Object {
            "alignment": "right",
            "content": [Function],
            "name": "createAt",
            "size": "shrink",
            "title": "created at",
          },
        ]
      }
      expandedSwipeCellRow={null}
      hasLeftAction={true}
      hasTableSwipeActions={false}
      interactionType="hover"
      isSelected={false}
      key="4"
      lastRow={true}
      row={
        Object {
          "actions": Array [
            Object {
              "icon": "link",
              "onClick": [Function],
              "tooltipText": "Copy Link",
              "type": "solid",
            },
            Object {
              "icon": "pencil-alt",
              "onClick": [Function],
              "tooltipText": "Rename",
              "type": "solid",
            },
            Object {
              "icon": "file-download",
              "onClick": [Function],
              "tooltipText": "Download",
              "type": "solid",
            },
            Object {
              "icon": "trash",
              "onClick": [Function],
              "tooltipText": "Delete",
              "type": "solid",
            },
          ],
          "data": Object {
            "createAt": "05/01/2018",
            "fileName": "test1.pdf",
            "fileType": "PDF",
            "size": "1.5",
          },
          "id": "4",
          "isSelectable": true,
        }
      }
      setExpandedSwipeCellRow={[Function]}
      toggleSelected={[Function]}
    />
  </styled.table>
</styled.div>
`;

exports[`<Table /> should render an empty table 1`] = `
<styled.div
  data-component-type="table"
>
  <styled.table>
    <styled.thead>
      <Memo(TableHeader)
        columns={
          Array [
            Object {
              "alignment": "left",
              "content": [Function],
              "name": "fileName",
              "size": "auto",
              "title": "file name",
            },
            Object {
              "alignment": "right",
              "content": [Function],
              "name": "createAt",
              "size": "shrink",
              "title": "created at",
            },
          ]
        }
        hasBulkAction={false}
        hasLeftAction={false}
        hasSortEnabled={true}
        hasTableSwipeActions={false}
        isEmpty={true}
        onSortChange={[MockFunction]}
        selectAllTableCheckboxInputValue="false"
        setSelectAllTableCheckboxInputValue={[Function]}
        sort={
          Object {
            "createAt": "ascending",
          }
        }
      />
    </styled.thead>
    <styled.tr>
      <styled.td
        colSpan={2}
      >
        <div>
          empty
        </div>
      </styled.td>
    </styled.tr>
  </styled.table>
</styled.div>
`;
